### PR TITLE
BaSP-TG-16:  Super admin seeds and tests

### PR DESCRIPTION
**Pull request by pabloschivazappa**

### CODE REVIEW

File: src/test/super-admins.test.js [hunk 1]
```diff
@@ -1,3 +1,139 @@
...
+      const response = await request(app).get('/super-admins').send();
+      expect(response.body.data.length).toBeGreaterThan(0);
+    });
+  });
+
+  describe('GETbyID /Super-admins', () => {
+    test('should GET a super admin by ID', async () => {
+      const response = await request(app).get('/super-admins').send();
+      expect(response.status).toBe(200);
+      expect(response.body.data).toBeDefined();
```
Comment (GinaSchP) [inline src/test/super-admins.test.js:43]: We should compare the data with the data that we have inside the seeds.

File: src/test/super-admins.test.js [hunk 2]
```diff
@@ -1,3 +1,139 @@
...
+
+  describe('POST /Super-admins', () => {
+    test('should create a super admin', async () => {
+      const response = await request(app).post('/super-admins').send(mockedSuperAdmin);
+      expect(response.status).toBe(201);
+      expect(response.body.error).toBeFalsy();
+      expect(response.body).toBeInstanceOf(Object);
+    });
+
+    test('should return error', async () => {
```
Comment (GinaSchP) [inline src/test/super-admins.test.js:62]:
```suggestion
   test('should return an error with status code 400 when we are trying to create an employee without body', async () => {
```

File: src/test/super-admins.test.js [hunk 3]
```diff
@@ -1,3 +1,139 @@
...
+      const response = await request(app).post('/super-admins').send();
+      expect(response.status).toBe(400);
+      expect(response.body.data).toBeUndefined();
+    });
```
Comment (GinaSchP) [inline src/test/super-admins.test.js:66]: We could add a new expect to check the error message and another to check if the error is true or false

File: src/test/super-admins.test.js [hunk 4]
```diff
@@ -1,3 +1,139 @@
...
+
+    test('the super admin should not be created, wrong name', async () => {
```
Comment (GinaSchP) [inline src/test/super-admins.test.js:68]:
```suggestion
    test('should return an error with status 400 when we are sending an invalid name in the request body', async () => {
```

File: src/test/super-admins.test.js [hunk 5]
```diff
@@ -1,3 +1,139 @@
...
+      const response = await request(app).post('/super-admins').send({
+        ...mockedSuperAdmin,
+        name: 'CS',
+      });
+      expect(response.status).toBe(400);
+      expect(response.body.error).toBeTruthy();
+    });
```
Comment (GinaSchP) [inline src/test/super-admins.test.js:75]: Add an expect to check the error message.

File: src/test/super-admins.test.js [hunk 6]
```diff
@@ -1,3 +1,139 @@
...
+
+    test('the super admin should not be created, wrong last name', async () => {
```
Comment (GinaSchP) [inline src/test/super-admins.test.js:77]: Same as the comments above.

File: src/test/super-admins.test.js [hunk 7]
```diff
@@ -1,3 +1,139 @@
...
+      const response = await request(app).put(`/super-admins/${validId}`).send({
+        ...mockedSuperAdmin,
+        name: 'NewName',
+      });
+      expect(response.status).toBe(200);
+      expect(response.body.error).toBeFalsy();
+      expect(response.body.data).toBeDefined();
+    });
+
+    test('should not edit a super admin', async () => {
```
Comment (lmalvasia) [inline src/test/super-admins.test.js:132]:
```suggestion
    test('should not edit a super admin when the user sends an invalid id', async () => {
```

File: src/test/super-admins.test.js [hunk 8]
```diff
@@ -1,3 +1,139 @@
...
+
+  describe('DELETE /Super-admins', () => {
+    test('should delete a super admin', async () => {
+      const response = await request(app).delete(`/super-admins/${validId}`).send();
+      expect(response.status).toBe(200);
+      expect(response.body.error).toBeFalsy();
+      expect(response.body.data).toBeDefined();
+    });
+
+    test('should not remove a super admin', async () => {
```
Comment (lmalvasia) [inline src/test/super-admins.test.js:113]:
```suggestion
    test('should not remove a super admin when the user sends an invalid id', async () => {
```

File: src/test/super-admins.test.js [hunk 9]
```diff
@@ -1,3 +1,139 @@
...
+    test('the super admin should not be created, wrong last name', async () => {
+      const response = await request(app).post('/super-admins').send({
+        ...mockedSuperAdmin,
+        lastName: 'GO',
+      });
+      expect(response.status).toBe(400);
+      expect(response.body.error).toBeTruthy();
+    });
+
+    test('the super admin should not be created, wrong e-mail', async () => {
```
Comment (GinaSchP) [inline src/test/super-admins.test.js:86]: Same as the comments above.

File: src/test/super-admins.test.js [hunk 10]
```diff
@@ -1,3 +1,139 @@
...
+      const response = await request(app).post('/super-admins').send({
+        ...mockedSuperAdmin,
+        email: 'mirage.com',
+      });
+      expect(response.status).toBe(400);
+      expect(response.body.error).toBeTruthy();
+    });
+
+    test('the super admin should not be created, wrong password', async () => {
```
Comment (GinaSchP) [inline src/test/super-admins.test.js:95]: Same as the comments above.

File: src/test/super-admins.test.js [hunk 11]
```diff
@@ -1,3 +1,139 @@
...
+    });
+
+    test('the super admin should not be created, wrong last name', async () => {
+      const response = await request(app).post('/super-admins').send({
+        ...mockedSuperAdmin,
+        lastName: 'GO',
+      });
+      expect(response.status).toBe(400);
+      expect(response.body.error).toBeTruthy();
+    });
```
Comment (GinaSchP) [inline src/test/super-admins.test.js:84]: Add an expect to check the error message.

File: src/test/super-admins.test.js [hunk 12]
```diff
@@ -1,3 +1,139 @@
...
+
+    test('the super admin should not be created, wrong e-mail', async () => {
+      const response = await request(app).post('/super-admins').send({
+        ...mockedSuperAdmin,
+        email: 'mirage.com',
+      });
+      expect(response.status).toBe(400);
+      expect(response.body.error).toBeTruthy();
+    });
```
Comment (GinaSchP) [inline src/test/super-admins.test.js:93]: Add an expect to check the error message.

File: src/test/super-admins.test.js [hunk 13]
```diff
@@ -1,3 +1,139 @@
...
+      expect(response.body.error).toBeFalsy();
+      expect(response.body.data).toBeDefined();
+    });
+
+    test('should not edit a super admin', async () => {
+      const response = await request(app).put(`/super-admins/${invalidID}`).send();
+      expect(response.status).toBe(400);
+      expect(response.body.error).toBeTruthy();
+      expect(response.body.data).toBeUndefined();
+    });
```
Comment (GinaSchP) [inline src/test/super-admins.test.js:137]: Add an expect to check the error message.

File: src/test/super-admins.test.js (hunk 3 above)
Comment (pabloschivazappa) [reply src/test/super-admins.test.js:66]: In the controller I have the message like this:       
`message: `An error occurred: ${error}`,`
And I don´t have the variable 'error' in test file.
Should I change the message in controller?

File: src/test/super-admins.test.js [hunk 14]
```diff
@@ -1,3 +1,154 @@
...
+  });
+
+  describe('GETbyID /Super-admins', () => {
+    test('should return a super admin by ID', async () => {
+      const response = await request(app).get(`/super-admins/${validId}`).send();
+      expect(response.status).toBe(200);
+      expect(response.body.data).toBeDefined();
+    });
+
+    test('should not return a super admin by ID', async () => {
```
Comment (DamianPalavecino) [inline src/test/super-admins.test.js:47]:
```suggestion
    test('should return an error when the user sends an invalid id', async () => {
```

File: src/test/super-admins.test.js [hunk 15]
```diff
@@ -1,3 +1,154 @@
...
+      const response = await request(app).get(`/super-admins/${invalidID}`).send();
+      expect(response.status).toBe(400);
+      expect(response.body.error).toBeTruthy();
+      expect(response.body.data).toBeUndefined();
+    });
+    test('should not return a super admin, ID not found', async () => {
+      const response = await request(app).get(`/super-admins/${IdNotFound}`).send();
+      expect(response.status).toBe(404);
+      expect(response.body.error).toBeTruthy();
+      expect(response.body.data).toBeNull();
```
Comment (DamianPalavecino) [inline src/test/super-admins.test.js:57]: Here you can add a new expect to check if the error message matches the one in the controller ('Super Admin not found')

File: src/test/super-admins.test.js [hunk 16]
```diff
@@ -1,3 +1,154 @@
...
+      expect(response.body.error).toBeFalsy();
+      expect(response.body.data).toBeDefined();
+    });
+
+    test('should not edit a super admin when the user sends an invalid id', async () => {
+      const response = await request(app).put(`/super-admins/${invalidID}`).send();
+      expect(response.status).toBe(400);
+      expect(response.body.error).toBeTruthy();
+      expect(response.body.data).toBeUndefined();
+      expect(response.body.message).toEqual('There was an error: "name" is required');
```
Comment (DamianPalavecino) [inline src/test/super-admins.test.js:151]: Here you can make two separated tests: One to test what happens when we send an invalid id and another to check what happens when we send an invalid body

File: src/test/super-admins.test.js [hunk 17]
```diff
@@ -1,3 +1,154 @@
...
+
+  // eslint-disable-next-line no-underscore-dangle
+  const validId = superAdminSeeds[0]._id;
+  const invalidID = '6349bc420340abc705a7e8a5naid';
+  const IdNotFound = '6349bc420340abc705a72222';
+
+  describe('GET /Super-admins', () => {
+    test('should return status code 200', async () => {
+      const response = await request(app).get('/super-admins').send();
+      expect(response.status).toBe(200);
```
Comment (DamianPalavecino) [inline src/test/super-admins.test.js:26]:
```suggestion
      expect(response.status).toBe(200);
      expect(response.body.error).toBeFalsy();
      expect(response.body.data.length).toBeGreaterThan(0);
```

Making two new separated tests for those two expects is a bit unnecessary and both of them can be inserted in the first test

Comment (lmalvasia) [reply src/test/super-admins.test.js:26]: I want to see more comments from you guys!! Thanks @DamianPalavecino 👏

File: src/test/super-admins.test.js [hunk 18]
```diff
@@ -1,3 +1,158 @@
...
+      expect(response.body.data).toBeDefined();
+    });
+
+    test('should return an error when the user sends an invalid id', async () => {
+      const response = await request(app).get(`/super-admins/${invalidID}`).send();
+      expect(response.status).toBe(400);
+      expect(response.body.error).toBeTruthy();
+      expect(response.body.data).toBeUndefined();
+    });
```
Comment (lmalvasia) [inline src/test/super-admins.test.js:44]: Add a new line here to separate the tests

File: src/test/super-admins.test.js [hunk 19]
```diff
@@ -1,3 +1,158 @@
...
+      expect(response.body.error).toBeFalsy();
+      expect(response.body).toBeInstanceOf(Object);
+    });
+
+    test('should return error', async () => {
+      const response = await request(app).post('/super-admins').send();
+      expect(response.status).toBe(400);
+      expect(response.body.data).toBeUndefined();
+      expect(response.body.error).toBeTruthy();
+      expect(response.body.mesage).toBeUndefined();
```
Comment (lmalvasia) [inline src/test/super-admins.test.js:67]:
```suggestion
      expect(response.body.message).toMatch('An error occurred');
```
or
```suggestion
      expect(response.body.mesage).toBeDefined();
```

File: src/test/super-admins.test.js (hunk 3 above)
Comment (lmalvasia) [reply src/test/super-admins.test.js:68]: Check the comment above.

File: src/test/super-admins.test.js (hunk 2 above)
Comment (lmalvasia) [reply src/test/super-admins.test.js:63]: Change this message 😂

Comment (pabloschivazappa) [reply src/test/super-admins.test.js]: Done 💪🏻

**Review by lmalvasia — APPROVED** [review]

**Pull request merged**